apps/openmw/mwmechanics/aicombat.cpp

```diff
@@ -326,95 +284,66 @@ namespace MWMechanics
         if (currentAction.get())
             currentAction->getCombatRange(rangeAttack, rangeFollow);
 
-        // FIXME: consider moving this stuff to ActionWeapon::getCombatRange
         const ESM::Weapon *weapon = NULL;
-        MWMechanics::WeaponType weaptype = WeapType_None;
-        float weapRange = 1.0f;
+        MWMechanics::WeaponType weapType = MWMechanics::WeaponType::WeapType_None;
+        float spellSpeed = 1.f;
 
-        // Get weapon characteristics
-        if (actorClass.hasInventoryStore(actor))
+        if (dynamic_cast<ActionWeapon*>(currentAction.get()))
         {
-            // TODO: Check equipped weapon and equip a different one if we can't attack with it
-            // (e.g. no ammunition, or wrong type of ammunition equipped, etc. autoEquip is not very smart in this regard))
+            weapon = dynamic_cast<ActionWeapon*>(currentAction.get())->getWeapon();
 
-            //Get weapon speed and range
-            MWWorld::ContainerStoreIterator weaponSlot =
-                MWMechanics::getActiveWeapon(actorClass.getCreatureStats(actor), actorClass.getInventoryStore(actor), &weaptype);
+            if (weapon != NULL)
+                weapType = getWeaponType(static_cast<ESM::Weapon::Type>(weapon->mData.mType));
+            else
+                weapType = WeapType_HandToHand;
 
-            if (weaptype == WeapType_HandToHand)
-            {
-                static float fHandToHandReach =
-                    world->getStore().get<ESM::GameSetting>().find("fHandToHandReach")->getFloat();
-                weapRange = fHandToHandReach;
-            }
-            else if (weaptype != WeapType_PickProbe && weaptype != WeapType_Spell && weaptype != WeapType_None)
+            if ((weapon == NULL && storage.mLastWeapon != "handToHand") 
+                || (weapon != NULL && weapon->mModel != storage.mLastWeapon))
             {
-                // All other WeapTypes are actually weapons, so get<ESM::Weapon> is safe.
-                weapon = weaponSlot->get<ESM::Weapon>()->mBase;
-                weapRange = weapon->mData.mReach;
+                storage.updateWeaponAttackDurations(actor, weapType, weapon == NULL ? 1.f : weapon->mData.mSpeed, ESM::RT_Self);
```

Comment:
> ...startCombat just contructs the AiPackage, but doesn't execute it

Anyway there will be rather unpleasant extension of AiSequence::execute, where you'll need to check if package is AiCombat and perform search for that CharacterController on every execution, so specifically for this maybe better would be to queue world state to find CharacterController in Aicombat constructor.
For bug (**#1876**) we can add whole animation duration, but also we can assume the follow part is not that significant in terms of time and just naively increase a period between them.
I agree that in terms of flexibility your approach is better, but those knockouts and blocks affect actual combat minimally.
You said previously
>That would get rid of all this duplicated logic, and make the AI code much simpler.

We anyway will need to write a somewhat similar to **updateWeaponAttackDurations** function which should decide if weapon group's animation is currently playing and if yes: check if it is min-to-max attack (and check complete % here) or pre-/post- min-to-max attack; if no - than increment timer till  next attack start. Also handle here if we are unconsious or blocking etc.

So IMO the only direct reason to move to CharacterController as of now - is adding random delay between actual end of attack animation and start of a new one. And there is more simple solution (but no so exact).